### DISCUSSION TITLE
Problem with the HID Keyboard Scancode sorting?

### DISCUSSION THREAD

**Discussion by harun27** [hathach/tinyusb#1551] [Q&A]
Hi,

I was playing around with the usb host_cdc_msc_hid example on my rpi pico and wanted to output the raw keyboard report. So I replaced the `process_kbd_report` function with this:

```
static void process_kbd_report(hid_keyboard_report_t const *report)
{
  for (uint8_t k = 0; k < 8; k++)
  {
    if (k == 0)
    {
      uart_putc(UART_ID, report->modifier);
    }
    else if (k == 1)
    {
      uart_putc(UART_ID, report->reserved);
    }
    else
    {
      uart_putc(UART_ID, report->keycode[k - 2]);
    }
  }

  fflush(stdout);
}
```

When I attach my keyboard to my pico, and analyze the UART signal, I get something weird. The general structure works, so there is firstly modifier byte, then the reserved byte, then there are 6 bytes of the keycode. The problem is the order of the keycodes. So if I press f.e. a, b and c in any order, I always get the keycode 0x06 0x05 0x04 0x00 0x00 0x00. As far as I know, the order in which we press the keys is important. So if I would hold firstly a, then c, then b, I should have get 0x04 0x06 0x05 0x00 0x00 0x00 or? That also happens with other keys. I thought like it firstly sorts the keycodes according to their size, but that size sorting doesn't apply for every keycode. It is just that some bytes have like more priority or so?

Since I am not an expert in USB or very familiar with this library, I was not sure whether this is actually a bug, or is it me using the library wrong?

**Comment by hathach** [comment]
when in doubt, check the usb specs

```
The order of keycodes in array fields has no significance. Order determination
is done by the host software comparing the contents of the previous report to
the current report. If two or more keys are reported in one report, their order is
indeterminate. Keyboards may buffer events that would have otherwise
resulted in multiple event in a single report.
```

![image](https://user-images.githubusercontent.com/249515/179471601-8040efcd-0bec-4438-9ea4-8c163cb43437.png)
